File: frontend/src/api/scheduled-tasks.ts
/**
 * API client for Scheduled Tasks management (Celery Beat)
 */

import { apiClient } from './client';

// Types matching backend models
export interface ScheduledTask {
  id: string;
  name: string;
  task_name: string;
  description?: string;
  schedule_type: 'cron' | 'interval';

  // Cron fields
  cron_minute?: string;
  cron_hour?: string;
  cron_day_of_week?: string;
  cron_day_of_month?: string;
  cron_month_of_year?: string;

  // Interval fields
  interval_value?: number;
  interval_unit?: 'seconds' | 'minutes' | 'hours' | 'days';

  // Task arguments
  args: any[];
  kwargs: Record<string, any>;
  queue?: string;

  // Status
  is_active: boolean;
  is_paused: boolean;
  total_run_count: number;
  last_run_at?: string;
  last_run_success?: boolean;
  last_run_error?: string;

  // Metadata
  created_at: string;
  updated_at: string;
  deleted_at?: string;
  created_by?: string;
  updated_by?: string;
}

export interface ScheduledTaskCreate {
  name: string;
  task_name: string;
  description?: string;
  schedule_type: 'cron' | 'interval';

  // Cron fields
  cron_minute?: string;
  cron_hour?: string;
  cron_day_of_week?: string;
  cron_day_of_month?: string;
  cron_month_of_year?: string;

  // Interval fields
  interval_value?: number;
  interval_unit?: 'seconds' | 'minutes' | 'hours' | 'days';

  // Task arguments
  args?: any[];
  kwargs?: Record<string, any>;
  queue?: string;
  is_active?: boolean;
}

export interface ScheduledTaskUpdate {
  name?: string;
  task_name?: string;
  description?: string;
  schedule_type?: 'cron' | 'interval';

  // Cron fields
  cron_minute?: string;
  cron_hour?: string;
  cron_day_of_week?: string;
  cron_day_of_month?: string;
  cron_month_of_year?: string;

  // Interval fields
  interval_value?: number;
  interval_unit?: 'seconds' | 'minutes' | 'hours' | 'days';

  // Task arguments
  args?: any[];
  kwargs?: Record<string, any>;
  queue?: string;
  is_active?: boolean;
  is_paused?: boolean;
}

export interface ScheduledTasksResponse {
  count: number;
  data: ScheduledTask[];
}

/**
 * List all scheduled tasks
 */
export async function listScheduledTasks(params?: {
  skip?: number;
  limit?: number;
  include_inactive?: boolean;
}): Promise<ScheduledTasksResponse> {
  const searchParams = new URLSearchParams();
  if (params?.skip !== undefined) searchParams.append('skip', params.skip.toString());
  if (params?.limit !== undefined) searchParams.append('limit', params.limit.toString());
  if (params?.include_inactive !== undefined) searchParams.append('include_inactive', params.include_inactive.toString());

  const query = searchParams.toString() ? `?${searchParams.toString()}` : '';
  const response = await apiClient.get(`/scheduled-tasks${query}`);
  return response.data;
}

/**
 * Get a specific scheduled task by ID
 */
export async function getScheduledTask(taskId: string): Promise<ScheduledTask> {
  const response = await apiClient.get(`/scheduled-tasks/${taskId}`);
  return response.data;
}

/**
 * Create a new scheduled task
 */
export async function createScheduledTask(
  data: ScheduledTaskCreate
): Promise<ScheduledTask> {
  const response = await apiClient.post('/scheduled-tasks', data);
  return response.data;
}

/**
 * Update an existing scheduled task
 */
export async function updateScheduledTask(
  taskId: string,
  data: ScheduledTaskUpdate
): Promise<ScheduledTask> {
  const response = await apiClient.patch(`/scheduled-tasks/${taskId}`, data);
  return response.data;
}

/**
 * Delete a scheduled task (soft delete)
 */
export async function deleteScheduledTask(taskId: string): Promise<void> {
  await apiClient.delete(`/scheduled-tasks/${taskId}`);
}

/**
 * Pause a scheduled task
 */
export async function pauseScheduledTask(taskId: string): Promise<ScheduledTask> {
  const response = await apiClient.post(`/scheduled-tasks/${taskId}/pause`);
  return response.data;
}

/**
 * Resume a paused scheduled task
 */
export async function resumeScheduledTask(taskId: string): Promise<ScheduledTask> {
  const response = await apiClient.post(`/scheduled-tasks/${taskId}/resume`);
  return response.data;
}

/**
 * Execute a scheduled task immediately (without waiting for schedule)
 */
export async function runTaskNow(taskId: string): Promise<{
  success: boolean;
  message: string;
  task_id: string;
}> {
  const response = await apiClient.post(`/scheduled-tasks/${taskId}/run-now`);
  return response.data;
}

/**
 * Get list of available Celery tasks (registered in the system + modules)
 */
export async function getAvailableCeleryTasks(): Promise<string[]> {
  try {
    const response = await apiClient.get('/scheduled-tasks/available-tasks');
    return response.data;
  } catch (error) {
    console.error('Failed to fetch available tasks:', error);
    // Fallback to core tasks if API call fails
    return [
      'app.tasks.send_email',
      'app.tasks.cleanup_old_files',
      'app.tasks.collect_stats',
      'app.tasks.execute_scheduled_backups',
      'app.tasks.generate_report',
      'app.tasks.sync_external_data',
      'app.tasks.process_queue_items',
      'app.tasks.update_cache',
    ];
  }
}

/**
 * Helper: Parse cron expression to human-readable format
 */
export function cronToHumanReadable(task: ScheduledTask): string {
  if (task.schedule_type === 'interval') {
    const value = task.interval_value || 1;
    const unit = task.interval_unit || 'minutes';
    return `Every ${value} ${unit}`;
  }

  const minute = task.cron_minute || '*';
  const hour = task.cron_hour || '*';
  const dayOfWeek = task.cron_day_of_week || '*';
  const dayOfMonth = task.cron_day_of_month || '*';
  const month = task.cron_month_of_year || '*';

  // Simple patterns
  if (minute === '*' && hour === '*' && dayOfWeek === '*' && dayOfMonth === '*' && month === '*') {
    return 'Every minute';
  }
  if (minute !== '*' && hour === '*' && dayOfWeek === '*' && dayOfMonth === '*' && month === '*') {
    return `Every hour at minute ${minute}`;
  }
  if (minute !== '*' && hour !== '*' && dayOfWeek === '*' && dayOfMonth === '*' && month === '*') {
    return `Daily at ${hour}:${minute.padStart(2, '0')}`;
  }
  if (minute !== '*' && hour !== '*' && dayOfWeek !== '*' && dayOfMonth === '*' && month === '*') {
    const days = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];
    const dayName = days[parseInt(dayOfWeek)] || `day ${dayOfWeek}`;
    return `Weekly on ${dayName} at ${hour}:${minute.padStart(2, '0')}`;
  }

  // Full cron expression
  return `${minute} ${hour} ${dayOfMonth} ${month} ${dayOfWeek}`;
}

/**
 * Helper: Validate cron expression component
 */
export function validateCronField(value: string, min: number, max: number): boolean {
  if (value === '*') return true;

  // Handle ranges (e.g., "1-5")
  if (value.includes('-')) {
    const [start, end] = value.split('-').map(Number);
    return start >= min && end <= max && start <= end;
  }

  // Handle lists (e.g., "1,3,5")
  if (value.includes(',')) {
    const values = value.split(',').map(Number);
    return values.every(v => v >= min && v <= max);
  }

  // Handle steps (e.g., "*/5")
  if (value.includes('/')) {
    const [range, step] = value.split('/');
    const stepNum = Number(step);
    if (range === '*') {
      return stepNum > 0 && stepNum <= max;
    }
  }

  // Simple number
  const num = Number(value);
  return num >= min && num <= max;
}

File: backend/app/api/routes/scheduled_tasks.py
"""
Routes API pour la gestion des tâches planifiées (Celery Beat)
"""

from typing import Any, List
from uuid import UUID
from fastapi import APIRouter, HTTPException, Depends
from sqlmodel import select, func

from app.api.deps import CurrentUser, SessionDep
from app.models_scheduled_tasks import (
    ScheduledTask,
    ScheduledTaskCreate,
    ScheduledTaskUpdate,
    ScheduledTaskPublic,
    ScheduledTasksResponse,
    TaskExecutionLog,
    TaskExecutionLogPublic,
    TaskExecutionLogsResponse,
)
from app.core.rbac import require_permission
from app.core.hook_trigger_service import hook_trigger
from datetime import datetime

router = APIRouter(prefix="/scheduled-tasks", tags=["scheduled-tasks"])


@router.get("")
@require_permission("core.scheduled_tasks.read")
async def list_scheduled_tasks(
    session: SessionDep,
    current_user: CurrentUser,
    skip: int = 0,
    limit: int = 100,
    include_inactive: bool = False,
) -> ScheduledTasksResponse:
    """
    Liste toutes les tâches planifiées.

    Requiert la permission: core.scheduled_tasks.read
    """
    statement = select(ScheduledTask).where(ScheduledTask.deleted_at == None)

    if not include_inactive:
        statement = statement.where(ScheduledTask.is_active == True)

    statement = statement.offset(skip).limit(limit)

    tasks = session.exec(statement).all()

    # Count total
    count_statement = select(func.count()).select_from(ScheduledTask).where(ScheduledTask.deleted_at == None)
    if not include_inactive:
        count_statement = count_statement.where(ScheduledTask.is_active == True)

    total = session.exec(count_statement).one()

    return ScheduledTasksResponse(
        count=total,
        data=[ScheduledTaskPublic.model_validate(task) for task in tasks]
    )


@router.get("/available-tasks")
@require_permission("core.scheduled_tasks.read")
async def get_available_tasks(
    session: SessionDep,
    current_user: CurrentUser,
) -> List[str]:
    """
    Récupère la liste des tâches Celery disponibles (CORE + modules actifs).

    Requiert la permission: core.scheduled_tasks.read

    Returns:
        Liste des noms de tâches disponibles triée par ordre alphabétique
    """
    from app.core.queue_service import celery_app
    from app.core.module_loader import ModuleLoader
    from app.models_modules import Module, ModuleStatus

    try:
        tasks = []

        # Récupérer les tâches réellement enregistrées dans les workers Celery
        inspector = celery_app.control.inspect()
        registered_tasks_by_worker = inspector.registered()

        if registered_tasks_by_worker:
            # Collecter toutes les tâches de tous les workers
            for worker_tasks in registered_tasks_by_worker.values():
                tasks.extend([
                    name for name in worker_tasks
                    if not name.startswith('celery.')
                ])

        # Si aucun worker n'est disponible, fallback sur les tâches CORE connues
        if not tasks:
            tasks = [
                'app.tasks.send_email',
                'app.tasks.cleanup_old_files',
                'app.tasks.collect_stats',
                'app.tasks.execute_scheduled_backups',
            ]

        # Charger les tâches des modules actifs
        statement = select(Module).where(
            Module.status == ModuleStatus.ACTIVE,
            Module.deleted_at == None
        )
        active_modules = session.exec(statement).all()

        for module in active_modules:
            module_tasks = ModuleLoader.load_module_tasks(module.code)
            tasks.extend(module_tasks)

        return sorted(list(set(tasks)))  # Dédupliquer et trier

    except Exception as e:
        import logging
        logging.getLogger(__name__).error(f"Failed to load available tasks: {e}")
        # Fallback: tâches CORE minimales
        return [
            'app.tasks.send_email',
            'app.tasks.cleanup_old_files',
            'app.tasks.collect_stats',
            'app.tasks.execute_scheduled_backups',
        ]


@router.get("/{task_id}")
@require_permission("core.scheduled_tasks.read")
async def get_scheduled_task(
    task_id: UUID,
    session: SessionDep,
    current_user: CurrentUser,
) -> ScheduledTaskPublic:
    """
    Récupère une tâche planifiée par son ID.

    Requiert la permission: core.scheduled_tasks.read
    """
    task = session.get(ScheduledTask, task_id)

    if not task or task.deleted_at:
        raise HTTPException(status_code=404, detail="Scheduled task not found")

    return ScheduledTaskPublic.model_validate(task)


@router.post("")
@require_permission("core.scheduled_tasks.create")
async def create_scheduled_task(
    task_data: ScheduledTaskCreate,
    session: SessionDep,
    current_user: CurrentUser,
) -> ScheduledTaskPublic:
    """
    Crée une nouvelle tâche planifiée.

    Requiert la permission: core.scheduled_tasks.create
    """
    # Check if name already exists
    existing = session.exec(
        select(ScheduledTask).where(
            ScheduledTask.name == task_data.name,
            ScheduledTask.deleted_at == None
        )
    ).first()

    if existing:
        raise HTTPException(status_code=400, detail="A task with this name already exists")

    # Validate schedule configuration
    if task_data.schedule_type == "interval" and not task_data.interval_value:
        raise HTTPException(status_code=400, detail="interval_value is required for interval schedule")

    task = ScheduledTask.model_validate(
        task_data.model_dump(),
        update={
            "created_by": current_user.id,
            "updated_by": current_user.id,
        }
    )

    session.add(task)
    session.commit()
    session.refresh(task)

    # Trigger hook
    try:
        await hook_trigger.trigger_event(
            event="scheduled_task.created",
            context={
                "task_id": str(task.id),
                "task_name": task.task_name,
                "created_by": str(current_user.id),
            },
            db=session,
        )
    except Exception as e:
        import logging
        logging.getLogger(__name__).warning(f"Failed to trigger scheduled_task.created hook: {e}")

    # Reload Celery Beat schedule
    from app.core.celery_beat_loader import reload_beat_schedule
    reload_beat_schedule()

    return ScheduledTaskPublic.model_validate(task)


@router.patch("/{task_id}")
@require_permission("core.scheduled_tasks.update")
async def update_scheduled_task(
    task_id: UUID,
    task_data: ScheduledTaskUpdate,
    session: SessionDep,
    current_user: CurrentUser,
) -> ScheduledTaskPublic:
    """
    Met à jour une tâche planifiée.

    Requiert la permission: core.scheduled_tasks.update
    """
    task = session.get(ScheduledTask, task_id)

    if not task or task.deleted_at:
        raise HTTPException(status_code=404, detail="Scheduled task not found")

    # Check name uniqueness if changing name
    if task_data.name and task_data.name != task.name:
        existing = session.exec(
            select(ScheduledTask).where(
                ScheduledTask.name == task_data.name,
                ScheduledTask.deleted_at == None,
                ScheduledTask.id != task_id
            )
        ).first()

        if existing:
            raise HTTPException(status_code=400, detail="A task with this name already exists")

    # Update fields
    update_data = task_data.model_dump(exclude_unset=True)
    for key, value in update_data.items():
        setattr(task, key, value)

    task.updated_at = datetime.utcnow()
    task.updated_by = current_user.id

    session.add(task)
    session.commit()
    session.refresh(task)

    # Trigger hook
    try:
        await hook_trigger.trigger_event(
            event="scheduled_task.updated",
            context={
                "task_id": str(task.id),
                "task_name": task.task_name,
                "updated_by": str(current_user.id),
            },
            db=session,
        )
    except Exception as e:
        import logging
        logging.getLogger(__name__).warning(f"Failed to trigger scheduled_task.updated hook: {e}")

    # Reload Celery Beat schedule
    from app.core.celery_beat_loader import reload_beat_schedule
    reload_beat_schedule()

    return ScheduledTaskPublic.model_validate(task)


@router.delete("/{task_id}")
@require_permission("core.scheduled_tasks.delete")
async def delete_scheduled_task(
    task_id: UUID,
    session: SessionDep,
    current_user: CurrentUser,
) -> dict:
    """
    Supprime (soft delete) une tâche planifiée.

    Requiert la permission: core.scheduled_tasks.delete
    """
    task = session.get(ScheduledTask, task_id)

    if not task or task.deleted_at:
        raise HTTPException(status_code=404, detail="Scheduled task not found")

    task.deleted_at = datetime.utcnow()
    task.is_active = False
    task.updated_by = current_user.id

    session.add(task)
    session.commit()

    # Trigger hook
    try:
        await hook_trigger.trigger_event(
            event="scheduled_task.deleted",
            context={
                "task_id": str(task.id),
                "task_name": task.task_name,
                "deleted_by": str(current_user.id),
            },
            db=session,
        )
    except Exception as e:
        import logging
        logging.getLogger(__name__).warning(f"Failed to trigger scheduled_task.deleted hook: {e}")

    # Reload Celery Beat schedule
    from app.core.celery_beat_loader import reload_beat_schedule
    reload_beat_schedule()

    return {"success": True, "message": "Scheduled task deleted"}


@router.post("/{task_id}/pause")
@require_permission("core.scheduled_tasks.update")
async def pause_scheduled_task(
    task_id: UUID,
    session: SessionDep,
    current_user: CurrentUser,
) -> ScheduledTaskPublic:
    """
    Met en pause une tâche planifiée.

    Requiert la permission: core.scheduled_tasks.update
    """
    task = session.get(ScheduledTask, task_id)

    if not task or task.deleted_at:
        raise HTTPException(status_code=404, detail="Scheduled task not found")

    task.is_paused = True
    task.updated_at = datetime.utcnow()
    task.updated_by = current_user.id

    session.add(task)
    session.commit()
    session.refresh(task)

    # Reload Celery Beat schedule
    from app.core.celery_beat_loader import reload_beat_schedule
    reload_beat_schedule()

    return ScheduledTaskPublic.model_validate(task)


@router.post("/{task_id}/resume")
@require_permission("core.scheduled_tasks.update")
async def resume_scheduled_task(
    task_id: UUID,
    session: SessionDep,
    current_user: CurrentUser,
) -> ScheduledTaskPublic:
    """
    Reprend l'exécution d'une tâche planifiée en pause.

    Requiert la permission: core.scheduled_tasks.update
    """
    task = session.get(ScheduledTask, task_id)

    if not task or task.deleted_at:
        raise HTTPException(status_code=404, detail="Scheduled task not found")

    task.is_paused = False
    task.updated_at = datetime.utcnow()
    task.updated_by = current_user.id

    session.add(task)
    session.commit()
    session.refresh(task)

    # Reload Celery Beat schedule
    from app.core.celery_beat_loader import reload_beat_schedule
    reload_beat_schedule()

    return ScheduledTaskPublic.model_validate(task)


@router.post("/{task_id}/run-now")
@require_permission("core.scheduled_tasks.execute")
async def run_task_now(
    task_id: UUID,
    session: SessionDep,
    current_user: CurrentUser,
) -> dict:
    """
    Exécute immédiatement une tâche planifiée (sans attendre le schedule).

    Requiert la permission: core.scheduled_tasks.execute
    """
    task = session.get(ScheduledTask, task_id)

    if not task or task.deleted_at:
        raise HTTPException(status_code=404, detail="Scheduled task not found")

    if not task.is_active:
        raise HTTPException(status_code=400, detail="Cannot run inactive task")

    # Enqueue the task immediately
    from app.core.queue_service import celery_app

    try:
        result = celery_app.send_task(
            task.task_name,
            args=task.args,
            kwargs=task.kwargs,
            queue=task.queue,
        )

        # Create execution log
        log = TaskExecutionLog(
            task_id=task_id,
            celery_task_id=result.id,
            started_at=datetime.utcnow(),
            status="pending",
        )
        session.add(log)
        session.commit()

        return {
            "success": True,
            "message": "Task enqueued for immediate execution",
            "celery_task_id": result.id,
            "log_id": str(log.id)
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to enqueue task: {str(e)}")


@router.get("/{task_id}/logs")
@require_permission("core.scheduled_tasks.read")
async def get_task_logs(
    task_id: UUID,
    session: SessionDep,
    current_user: CurrentUser,
    skip: int = 0,
    limit: int = 50,
) -> TaskExecutionLogsResponse:
    """
    Récupère l'historique d'exécution d'une tâche planifiée.

    Requiert la permission: core.scheduled_tasks.read
    """
    # Verify task exists
    task = session.get(ScheduledTask, task_id)
    if not task or task.deleted_at:
        raise HTTPException(status_code=404, detail="Scheduled task not found")

    # Get logs
    statement = (
        select(TaskExecutionLog)
        .where(TaskExecutionLog.task_id == task_id)
        .order_by(TaskExecutionLog.started_at.desc())
        .offset(skip)
        .limit(limit)
    )
    logs = session.exec(statement).all()

    # Count total
    count_statement = select(func.count()).select_from(TaskExecutionLog).where(
        TaskExecutionLog.task_id == task_id
    )
    total = session.exec(count_statement).one()

    return TaskExecutionLogsResponse(
        count=total,
        data=[TaskExecutionLogPublic.model_validate(log) for log in logs]
    )
